io-engine/src/bdev/nexus/nexus_io.rs

```diff
@@ -301,23 +300,74 @@ impl<'n> NexusBio<'n> {
                         IoSubmissionFailure::Write,
                     ),
                 );
-
-                self.fail();
+                r
             } else {
                 self.ctx_mut().in_flight = 1;
+                r
             }
-            r
         } else {
             error!(
                 "{:?}: read I/O submission failed: no children available",
                 self
             );
 
-            self.fail();
             Err(CoreError::NoDevicesAvailable {})
         }
     }
 
+    /// Submit a read operation to the next suitable replica.
+    /// In case of submission error the requiest is transparently resubmitted
+    /// to the next available replica.
+    fn do_readv(&mut self) -> Result<(), CoreError> {
+        match self.__do_readv_one() {
+            Err(e) => {
+                match e {
+                    // No readers available - bail out.
+                    CoreError::NoDevicesAvailable {} => {
+                        self.fail();
+                        Err(e)
+                    }
+                    // Failed to submit Read I/O request to the current replica,
+                    // try to resumbit request to the next available replica.
+                    _ => {
+                        let mut num_readers = self.channel().num_readers();
+
+                        let r = {
+                            if num_readers <= 1 {
+                                // No more readers available (taking into
+                                // account the failed one).
+                                Err(e)
+                            } else {
+                                num_readers -= 1; // Account the failed reader.
+
+                                // Resubmission loop to find a next available
+                                // replica for this Read I/O operation.
+                                loop {
+                                    match self.__do_readv_one() {
```

Comment:
How do we ensure it won't select the same reader again?

Reply:
Readers are selected based onround-robin policy. So we have to check N-1 remaining readers.